corehq/apps/users/management/commands/migrate_reprs.py

```diff
@@ -1,37 +1,36 @@
 import logging
 
 from django.core.management.base import BaseCommand
+from django.db.models import Q
 from corehq.apps.users.models import UserHistory
-from corehq.apps.users.util import user_id_to_username
+from corehq.apps.users.util import cached_user_id_to_username
 
 from corehq.apps.users.util import SYSTEM_USER_ID
 from corehq.util.log import with_progress_bar
 
-
 class Command(BaseCommand):
     help = "Adds user_repr and changed_by_repr to UserHistory log if not already present"
 
     def handle(self):
-        records = UserHistory.objects
-        for user_history in with_progress_bar(records.order_by('pk').iterator(), 185):
-            if user_history.changed_by_repr is None or user_history.user_repr is None:
-                try:
-                    migrate(user_history)
-                except Exception as e:
-                    logging.error(f"{user_history.pk}: {e}")
+        records = UserHistory.objects.filter(Q(user_repr__isnull=True) | Q(changed_by_repr__isnull=True))
+        for user_history in with_progress_bar(records.order_by('pk').iterator(), records.count()):
+            try:
+                migrate(user_history)
+            except Exception as e:
+                logging.error(f"{user_history.pk}: {e}")
 
 
 def migrate(user_history):
     if user_history.changed_by == SYSTEM_USER_ID:
         changed_by_repr = SYSTEM_USER_ID
     else:
-        changed_by_repr = user_id_to_username(user_history.changed_by)
+        changed_by_repr = cached_user_id_to_username(user_history.changed_by)
 
-    user_repr = user_id_to_username(user_history.user_id)
+    user_repr = cached_user_id_to_username(user_history.user_id)
 
-    if changed_by_repr:
-        user_history.change_by_repr = changed_by_repr
-    if user_repr:
+    if changed_by_repr and user_history.changed_by_repr != changed_by_repr:
```

Comment:
I meant that you should check whether there's a change before deciding whether to save - I don't think there's any cost to setting the variables, the save is the expensive part that's nice to skip.